paperqa/llms.py

```diff
@@ -0,0 +1,169 @@
+        and callbacks is a list of functions to call with each chunk of the completion.
+    """
+    if client is None:
+        raise ValueError(
+            "Your client is None - did you forget to set it after pickling?"
+        )
+    if llm_config["model_type"] == "chat":
+        system_message_prompt = dict(role="system", content=system_prompt)
+        human_message_prompt = dict(role="user", content=prompt)
+        if skip_system:
+            chat_prompt = [human_message_prompt]
+        else:
+            chat_prompt = [system_message_prompt, human_message_prompt]
+
+        async def execute(
+            data: dict, callbacks: list[Callable[[str], None]] | None = None
+        ) -> str:
+            messages = chat_prompt[:-1] + [
+                dict(role="user", content=chat_prompt[-1]["content"].format(**data))
+            ]
+            if callbacks is None:
+                completion = await client.chat.completions.create(
+                    messages=messages, **process_llm_config(llm_config)
+                )
+                output = completion.choices[0].message.content
+            else:
+                completion = await client.chat.completions.create(
+                    messages=messages, **process_llm_config(llm_config), stream=True
+                )
+                result = []
+                async for chunk in completion:
+                    c = chunk.choices[0].delta.content
+                    if c:
+                        result.append(c)
+                        [f(c) for f in callbacks]
+                output = "".join(result)
+            return output
+
+        return execute
+    elif llm_config["model_type"] == "completion":
+        if skip_system:
+            completion_prompt = prompt
+        else:
+            completion_prompt = system_prompt + "\n\n" + prompt
+
+        async def execute(
+            data: dict, callbacks: list[Callable[[str], None]] | None = None
+        ) -> str:
+            if callbacks is None:
+                completion = await client.completions.create(
+                    prompt=completion_prompt.format(**data),
+                    **process_llm_config(llm_config),
+                )
+                output = completion.choices[0].text
+            else:
+                completion = await client.completions.create(
+                    prompt=completion_prompt.format(**data),
+                    **process_llm_config(llm_config),
+                    stream=True,
+                )
+                result = []
+                async for chunk in completion:
+                    c = chunk.choices[0].text
+                    if c:
+                        result.append(c)
+                        [f(c) for f in callbacks]
+                output = "".join(result)
+            return output
+
+        return execute
+    else:
+        raise NotImplementedError(f"Unknown model type {llm_config['model_type']}")
+
+
+def get_score(text: str) -> int:
```

Comment:
I have a suggestion for making this logic more accessible (if I/chatgpt understands the current logic):
```

def get_score(text: str) -> int:
    # Check for N/A in the last line as a substring
    last_line = text.split("\n")[-1].lower()
    if "n/a" in last_line or "na" in last_line:
        return 0

    # Search for score patterns
    patterns = [
        r"[sS]core[:is\s]+([0-9]+)",  # Score: number
        r"\(([0-9])\w*\/",            # (number/
        r"([0-9]+)\w*\/"              # number/
    ]

    for pattern in patterns:
        match = re.search(pattern, text)
        if match:
            score = int(match.group(1))
            return min(score, 10) if score > 10 else score

    # Default scores based on text length
    return 1 if len(text) < 100 else 5
```

Reply:
This didn't pass unit tests - this function confuses me for sure. Will get back to this later though - maybe we can separate this from regex

Reply:
Looked more into this - let's wait until we have the json format thing we discussed implemented